runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,557 @@
+
+  local promise = require('vim.ui.img.utils.promise').new({
+    context = 'image.convert',
+  })
+
+  -- Schedule now to allow access to the ImageMagick option
+  -- in the case that we're currently in a fast function
+  vim.schedule(function()
+    ---@type string[]
+    local cmd = require('vim.ui.img.utils').split_quoted(vim.o.imgprg)
+    table.insert(cmd, 'convert')
+    table.insert(cmd, self.filename)
+
+    if opts.background then
+      table.insert(cmd, '-background')
+      table.insert(cmd, opts.background)
+      table.insert(cmd, '-flatten')
+    end
+    if opts.crop then
+      local region = opts.crop:to_pixels()
+      table.insert(cmd, '-crop')
+      table.insert(
+        cmd,
+        string.format('%sx%s+%s+%s', region.width, region.height, region.x, region.y)
+      )
+    end
+    if opts.size then
+      local size_px = opts.size:to_pixels()
+      table.insert(cmd, '-resize')
+      table.insert(cmd, string.format('%sx%s', size_px.width, size_px.height))
+    end
+
+    if opts.quality then
+      table.insert(cmd, '-quality')
+      table.insert(cmd, tostring(opts.quality))
+    end
+
+    local format = opts.format or 'png'
+    local output = opts.out or '-'
+    table.insert(cmd, format .. ':' .. output)
+
+    local ok, err = pcall(vim.system, cmd, nil, function(out)
+      if out.code ~= 0 then
+        promise:fail(out.stderr and out.stderr or 'failed to convert image')
+        return
+      end
+
+      -- In the case we wrote to a file, there's nothing to capture
+      if opts.out ~= nil then
+        promise:ok('')
+        return
+      end
+
+      local data = out.stdout
+      if not data or data == '' then
+        promise:fail('converted image output missing')
+        return
+      end
+
+      promise:ok(data)
+    end)
+
+    if not ok then
+      ---@cast err +string
+      ---@cast err -vim.SystemObj
+      promise:fail(err)
+    end
+  end)
+
+  return promise
+end
+
+---Displays an image, returning a reference to its visual representation (placement).
+---```lua
+---local img = ...
+---
+-----Can be invoked synchronously
+---local placement = assert(img:show({ ... }):wait())
+---
+-----Can also be invoked asynchronously
+---img:show({ ... }):on_done(function(err, placement)
+---  -- Do something
+---end)
+---```
+---@param opts? vim.ui.img.Opts|{provider?:string}
+---@return vim.ui.img.utils.Promise<vim.ui.img.Placement>
+function M:show(opts)
+  local promise = require('vim.ui.img.utils.promise').new({
+    context = 'image.show',
+  })
+
+  local placement = self:new_placement(opts)
+  placement
+    :show(opts)
+    :on_ok(function()
+      promise:ok(placement)
+    end)
+    :on_fail(function(err)
+      promise:fail(err)
+    end)
+
+  return promise
+end
+
+---Creates a placement of this image that is not yet visible.
+---@param opts? {provider?:string}
+---@return vim.ui.img.Placement
+function M:new_placement(opts)
+  return require('vim.ui.img.placement').new(self, opts)
```

Comment:
seems overwrought to represent "placements" as objects which need to be created. I would expect something like `M:set_pos`.

Positions are called "pos". Avoid a new "placement" concept.

Reply:
Maybe the naming is wrong on this - was following kitty's naming - but the concept is that the image is the file path and (optionally) data of the image itself and a placement is the reference to the image managed by the provider in the terminal (or UI) of which there may be multiple. Maybe we just don't make this a public method or we change the name, but `M:set_pos` makes me think that I'm moving the `vim.ui.Image` instead of one of the displayed versions on the terminal.

![neovim-img-placement-example](https://github.com/user-attachments/assets/fe57b401-cf0d-4be7-b868-8369ad9c92ba)

Reply:
> Maybe we just don't make this a public method or we change the name,

If it can be private for now, let's do that.

> but `M:set_pos` makes me think that I'm moving the `vim.ui.Image` instead of one of the displayed versions on the terminal.

Is the intention to avoid creating a new image for each placement? Can we instead just say "a new placement requires new image". So an image object has *one* position, and cannot be placed in multiple places. That might be "inefficient", but so what?

Reply:
The use cases for showing the same image multiple times, is probably quite limited. I can only really think of demo type effects, games and repeated UI elements. So, I think limiting the interface to only support images would be fine.

Furthermore, I'm pretty sure this can be optimized on the provider side. It can calculate the hash of the image data and properties and internally share it and keep track of the ids.

Reply:
> The use cases for showing the same image multiple times, is probably quite limited. I can only really think of demo type effects, games and repeated UI elements. So, I think limiting the interface to only support images would be fine.

Totally fair! I don't have a use case for same image repeated, and should be able to merge placement and image without much trouble. 

> Furthermore, I'm pretty sure this can be optimized on the provider side. It can calculate the hash of the image data and properties and internally share it and keep track of the ids.
> 
> 

Yep! I already do that within the sixel and iterm2 providers, and it works well. So agreed the optimization can be on the provider side. :smile: